Modularize tasks in hst

juzen2003 opened the pull request:
- Progress: https://github.com/orgs/SETI/projects/14/views/1
    - Work in progress items:
        - hst-helper (query_utils, fs_utils, more to be added)
        - tests for each module
        - finalize-hst-bundle
    - Pending items: (modules)
        - update-hst-program (only perform queue actions)
        - update-hst-visit (only perform queue actions)
        - task queue
    - Done items:
       - query-hst-moving-targets
       - query-hst-products
       - get-program-info
       - retrieve-hst-visit
       - label-hst-products (Modify Mark's code) 
       - prepare-browse-products
       - mark down doc

rfrenchseti commented on HST/citations/__init__.py:
```diff
@@ -151,10 +151,15 @@ def write(self, filename):
-            f.write(self.abstract)
+            f.write(self.title)
             f.write('\n')
+            if isinstance(self.abstract, list):
```
I think you only care if abstract is an iterable, right? So it could be a list, tuple, dictionary, whatever. In which case the modern way to do this is:
```
from collections.abs import Iterable
... isinstance(self.abstract, Iterable) ...
```

rfrenchseti commented on HST/get_program_info/__init__.py:
```diff
@@ -0,0 +1,133 @@
+
+def get_program_info(proposal_id, download_dir=None, logger=None):
+    """
+    Download proposal files and generate PROGRAM_INFO_FILE for the given proposal ID
```
Doc string should be on same line as """

rfrenchseti commented on HST/get_program_info/__init__.py:
```diff
@@ -0,0 +1,133 @@
+    if os.path.exists(filepath):
+        with open(filepath, 'rb') as f:
+            file_contents = f.read()
+        if file_contents == new_contents:
```
How about just `return file_contents != new_contents`

rfrenchseti commented on HST/hst_helper/__init__.py:
```diff
@@ -0,0 +1,28 @@
+           'bundles': os.environ['HST_BUNDLES']}
+HST_STAGING_DIR = os.environ['HST_STAGING']
+HST_PIPELINE_DIR = os.environ['HST_PIPELINE']
+HST_BUNDLES_DIR = os.environ['HST_BUNDLES']
```
Is there a reason to have both the dictionary and the global variables?

rfrenchseti commented on HST/hst_helper/__init__.py:
```diff
@@ -0,0 +1,28 @@
+HST_BUNDLES_DIR = os.environ['HST_BUNDLES']
+
+# suffixes for proposal files
+DOCUMENT_SUFFIXES = ['apt', 'pdf', 'pro', 'prop']
```
If these lists aren't going to be changed, make the tuples instead so they aren't mutable

rfrenchseti commented on HST/hst_helper/fs_utils.py:
```diff
@@ -0,0 +1,90 @@
+    """
+    program_dir = get_program_dir_path(proposal_id, visit, root_dir)
+
+    if not os.path.isdir(program_dir):
```
os.makedirs has an `exist_ok` parameter that removes the need for this test.

rfrenchseti commented on HST/hst_helper/fs_utils.py:
```diff
@@ -0,0 +1,90 @@
+    Input:
+        filepath:   the path of the targeted file
+    """
+    CHUNK = 4096
```
No real reason for this to be capitalized. Maybe chunk_size?

rfrenchseti commented on HST/hst_helper/query_utils.py:
```diff
@@ -0,0 +1,194 @@
+
+def ymd_tuple_to_mjd(ymd):
+    """Return Modified Julian Date.
+    Input:
```
In general these would be more readable if there were a blank line between the first line and the Input section

rfrenchseti commented on HST/pipeline/get-program-info.py:
```diff
@@ -0,0 +1,71 @@
+    help='The proposal id for the mast query')
+
+parser.add_argument('--log', '-l', type=str, default='',
+    help='Path and name for the log file. The name always has the current date and time '+
```
You can use """ multi-line strings here. Help ignores extra spaces so indenting the second+ lines doesn't actually show up as spaces in the help message.

rfrenchseti commented on HST/pipeline/query-hst-moving-targets.py:
```diff
@@ -0,0 +1,108 @@
+
+from query_hst_moving_targets import query_hst_moving_targets
+from hst_helper import (START_DATE,
+                         END_DATE,
```
Bad indent

rfrenchseti commented on HST/pipeline/query-hst-moving-targets.py:
```diff
@@ -0,0 +1,108 @@
+        logpath = parts[0] + '-' + now + parts[1]
+else:
+    if not os.path.isdir(LOG_DIR):
+        os.makedirs(LOG_DIR)
```
See earlier comment

rfrenchseti commented on HST/run_coverage.sh:
```diff
@@ -0,0 +1,2 @@
+coverage run -m pytest query_mast/tests
```
You should have a .coveragerc that sets up parameters and also turns on branch analysis. See OPUS or pds-webtools for examples.

rfrenchseti reviewed (changes requested):
I just did a quick look-through, not a detailed review.

- Remove .vscode from .gitignore and delete .vscode directory from GitHub
- All of the files seem to be missing a newline at the end of file. Maybe this is a VS Code setting?
- Some files use - and some use _ as separators. Be consistent.

-

rfrenchseti reviewed (approved):
I just did a basic overview of the code, not the functionality. Mark should check the functionality.

juzen2003 commented:
Address issues: #39, #40, #42, #44

rfrenchseti commented:
See my note in Slack about using the magic GitHub keywords to link this pull with those issues.

juzen2003 commented:
- Add more changes:
  - Implement these two tasks:
    - prepare-browse-products
    - label-hst-products (Modify Mark's code) 
   - Rename files under HST/pipeline to avoid name conflict with the corresponding task modules
   - Fixed descriptions of files under HST/pipeline

- Project progress: https://github.com/orgs/SETI/projects/14/views/1
- Fixes #39. Fixes #40. Fixes #42. Fixes #44. Fixes #45. Fixes #46.

juzen2003 commented:
Update:
- Add steps to generate the  document directory in the finalize_hst_bundle task.
- In ```finalize_document``` module, the ```label_hst_document_directory``` function perform the following tasks:
    - Create document directory.
    - Move/copy proposal files over from pipeline directory.
    - Create document label.
    - Create document collection csv.
    - Create document collection label.

juzen2003 commented:
Update:
- add step to create schema directory in the final bundle
- create schema collection csv & xml label files

Note:
finalize_hst_bundle task will now create document & schema directories. Next implementations in this task will be:
- create context directory
- move (or copy) products folders over
- create collection csv and xml for the products folders
- create bundle xml file

Example commands to run the current available tasks:
```
python pipeline/pipeline_query_hst_moving_targets.py -pid 7885
python pipeline/pipeline_query_hst_products.py -pid 7885
python pipeline/pipeline_get_program_info.py -pid 7885
python pipeline/pipeline_retrieve_hst_visit.py -pid 7885 -vi 01
python pipeline/pipeline_retrieve_hst_visit.py -pid 7885 -vi 02
python pipeline/pipeline_retrieve_hst_visit.py -pid 7885 -vi 03
python pipeline/pipeline_label_hst_products.py -pid 7885 -vi 01
python pipeline/pipeline_label_hst_products.py -pid 7885 -vi 02
python pipeline/pipeline_label_hst_products.py -pid 7885 -vi 03
python pipeline/pipeline_prepare_browse_products.py -pid 7885 -vi 01
python pipeline/pipeline_prepare_browse_products.py -pid 7885 -vi 02
python pipeline/pipeline_prepare_browse_products.py -pid 7885 -vi 03
python pipeline/pipeline_finalize_hst_bundle.py -pid 7885
```

The command to run current tests: (WIP, need to add more tests espeically once that generate xml label)
```
coverage run -m pytest
```

rfrenchseti commented:
Test comment

rfrenchseti commented:
Just to clarify, command line options usually don't have `_` in them. So maybe it should be `--prog-id`.

juzen2003 commented:
Update: Fixed #51 
- add step to create context directory in the final bundle
- create context collection csv & xml label files
- create investigation label

Note:
finalize_hst_bundle task will now create document, schema, and context directories. Next implementations in this task will be:
- move (or copy) products folders over
- create collection csv and xml for the products folders
- create bundle xml file

juzen2003 commented:
Update:
- Add these steps in finalize_hst_bundle task
  - Create collection csv for data/misc directories
  - Create collection label for data/misc directories
Note: finalize_hst_bundle will perform these
- Create document directory & collection csv and label
- Create schema directory & its collection csv and label
- Create context directory & its collection csv and label
- Move files from the staging directory to bundles directory
- Create collection csv and label for all data/misc directories
One final implementation left for this task is to create bundle label (WIP)

juzen2003 commented:
Update:
- Finalize hst bundle task done, these are the actions perform in the task:
  - Generate the final document directory and its csv & xml files
  - Generate the final schema directory and its csv & xml files
  - Generate the final context directory and its csv & xml files
  - Organize files, move all data files from staging to bundles
  - Create data collection csv & xml files
  - Create bundle label
  - Create manifest files & run validator

Note:
  - validate report (using the latest version 3.1.1) has some failures, looking into it now. A bundle of 7885 is uploaded to the dropbox under `/Shared-pdart/HST/bundles/hst_07885`

juzen2003 commented:
Update:
- Add a step to create context target label if it doesn't exist in the PDS page.
- Add tmp-context-products.json before running validation tool. 
- Fixed #54, Fixed #55, Fixed #47

markshowalter reviewed (changes requested):
so far

rfrenchseti commented on HST/finalize_context/__init__.py:
```diff
@@ -0,0 +1,114 @@
+        data_dict:      a data dictionary used to create the label.
+        logger:         pdslogger to use; None for default EasyLogger.
+    """
+    logger = logger or pdslogger.EasyLogger()
```
I actually prefer not to have the blank line after the multi-line docstring. PEP257 doesn't seem to specify a preference.

juzen2003 commented on HST/citations/__init__.py:
```diff
@@ -151,10 +152,15 @@ def write(self, filename):
-            f.write(', '.self.title)
-            f.write('\n')
-            f.write(self.abstract)
+            f.write(self.title)
```
Done

juzen2003 commented on HST/create_target_label/__init__.py:
```diff
@@ -0,0 +1,113 @@
+        logger.exception(ValueError)
+        raise ValueError(f'Proposal id: {proposal_id} is not valid.')
+
+    with urllib.request.urlopen(PDS_URL) as response:
```
Done

juzen2003 commented on HST/create_target_label/__init__.py:
```diff
@@ -0,0 +1,113 @@
+    soup = bs4.BeautifulSoup(html, "html.parser")
+    a_tags = soup.find_all("a")
+    pds_targ_list = [a.string for a in a_tags if a.string]
+    target_records = data_dict['target_identifications']
```
Done

juzen2003 commented on HST/finalize_context/__init__.py:
```diff
@@ -0,0 +1,114 @@
+    3. create context xml label.
+    4. create individual hst xml label.
+
+
```
Done

juzen2003 commented on HST/finalize_context/__init__.py:
```diff
@@ -0,0 +1,114 @@
+##########################################################################################
```
For now each task is structured like that and I plan to add task specific tests inside each module (subdirectory) so there will be tests subdirectory inside each task subdirectory. Once that part is implemented, if there is any subdirectories with only __init__.py later, I'll clean it up.

juzen2003 replied:
Restructure the code to have subdirectories with their own init.py files only when the directory is likely to contain additional files.

rfrenchseti commented on .vscode/settings.json:
```diff
@@ -0,0 +1,3 @@
+{
+    "python.analysis.extraPaths": ["./HST", "./HST"]
```
I don't think this needs to be here twice

rfrenchseti commented on HST/create_target_label.py:
```diff
@@ -0,0 +1,123 @@
+##########################################################################################
+# create_target_label/__init__.py
```
Need to update these comments now that they are not in subdirs

rfrenchseti commented on HST/create_target_label.py:
```diff
@@ -0,0 +1,123 @@
+    """Create target lable under context collection if it doesn't exist in PDS page.
+
+    Inputs:
+        proposal_id:    a proposal id.
```
We generally don't put colons after the parameter names in this kind of comment

rfrenchseti commented on HST/create_target_label.py:
```diff
@@ -0,0 +1,123 @@
+TARGET_LABEL_TEMPLATE = 'TARGET_LABEL.xml'
+
+def create_target_label(proposal_id, data_dict, logger=None):
+    """Create target lable under context collection if it doesn't exist in PDS page.
```
lable->label

rfrenchseti commented on HST/finalize_data_product.py:
```diff
@@ -0,0 +1,139 @@
+    for dir in os.listdir(deliverable_path):
+        for col_name in COL_NAME_PREFIX:
+            if dir.startswith(col_name): # work on data_directory
+                col_name = dir
```
It's not a great idea to change the variable of a for loop, because someone might try to use the original value later and cause a bug. I would name this variable something else.

rfrenchseti commented on HST/finalize_data_product.py:
```diff
@@ -0,0 +1,139 @@
+                            format_term = get_format_term(file)
+                            formatted_proposal_id = get_formatted_proposal_id(proposal_id)
+                            prod_lidvid = (f'P,urn:nasa:pds:hst_{formatted_proposal_id}'
+                                        + f':{dir}:{format_term}::'
```
You don't need the `+` here, and can indent the strings an extra space.

rfrenchseti commented on HST/finalize_document.py:
```diff
@@ -0,0 +1,163 @@
+    """With a given proposal id, create document directory in the final bundle. Return a
+    tupel of the path of document collection label and the path of document label. These
+    are the actions performed:
+    1. Create document directory.
```
I would find all these comments easier to read if there was a blank line between the first paragraph and the numbered list.

rfrenchseti commented on HST/finalize_document.py:
```diff
@@ -0,0 +1,163 @@
+            basename, _, ext = file.rpartition('.')
+            if ext in DOCUMENT_EXT or file == PROGRAM_INFO_FILE:
+                proposal_files_li.append((basename, file))
+                fp = pipeline_dir + f'/{file}'
```
Normally the variable `fp` means "file pointer" (although admittedly this is more of a C convention). For example: https://stackoverflow.com/questions/32049884/file-pointer-in-c

This code confused me for a little until I realized `fp` was a string. It might be nice to name it something else.

rfrenchseti replied:
For this and many other similar examples, it seems odd to have `string1 + f-string` when you could include `string1` in the f-string. For example:

```
fp = f'{pipeline_dir}/{file}'
```

I find this much more readable.

rfrenchseti commented on HST/finalize_document.py:
```diff
@@ -0,0 +1,163 @@
+        'records_num': records_num,
+        'mod_history': mod_history,
+    }
+    doc_data_dict = {**doc_data_dict, **data_dict}
```
Might be easier to say `doc_data_dict.update(data_dict)`

rfrenchseti commented on HST/finalize_document.py:
```diff
@@ -0,0 +1,163 @@
+    create_document_collection_csv(proposal_id, doc_data_dict, logger, testing)
+    # Create document collection label
+    doc_col_lbl = create_collection_label(proposal_id, 'document', doc_data_dict,
+                                   COL_DOC_LABEL, COL_DOC_LABEL_TEMPLATE,
```
Indent further

rfrenchseti commented on HST/finalize_hst_bundle.py:
```diff
@@ -0,0 +1,134 @@
+    # Get citation info
+    citation_info = get_citation_info(proposal_id, logger) if not testing else None
+    formatted_title = (citation_info.title
+    + ', HST Cycle '
```
Indent. Might be a lot nicer as one or more f-strings.

rfrenchseti commented on HST/finalize_hst_bundle.py:
```diff
@@ -0,0 +1,134 @@
+    timetag = os.path.getmtime(__file__)
+    label_date = datetime.datetime.fromtimestamp(timetag).strftime('%Y-%m-%d')
+
+    # Get imstrument id
```
instrument

rfrenchseti commented on HST/get_program_info.py:
```diff
@@ -0,0 +1,142 @@
+    """Download proposal files and generate PROGRAM_INFO_FILE for the given proposal ID.
+
+    Input:
+        proposal_id:    a proposal id.
```
Any function that returns a value should also have a "Returns:" section in the docstring.

rfrenchseti replied:
Any function that returns a value should also have a "Returns:" section in the docstring. (This is true throughout the code base)

rfrenchseti commented on HST/query_hst_products.py:
```diff
@@ -0,0 +1,225 @@
+                               f'{fname}', checksum_included)
+    else:
+        # Generate txt file if it doesn't exist
+        generate_files_txt(proposal_id, files_dict, visit, f'{fname}', checksum_included)
```
No point in `f'{fname}`, just say `fname`.

rfrenchseti reviewed (changes requested):
Just a few initial comments as I scan the code. I'll make comments in batches. Many of these comments apply to the entire code base, but I only mentioned them once.

juzen2003 commented:
Update based on Rob's feedback.

juzen2003 commented:
- Comments from Mark:
"As I mentioned yesterday, I am comfortable not reviewing the code line-by-line and instead evaluating the products it produces. The online samples that you've generated so far look great. I see one thing that puzzled me--the collections are called either "miscellaneous*" or "browse*"; there are no "data* collections. Furthermore, the browse collections contain FITS files, where I would expect them to contain PNGs or JPEGs. Is it possible that these collections are really "data" collections are they are mis-named "browse"? If so, that ought to be an easy fix.
Beyond that, I do recommend that you merge your changes into the main branch and proceed to run more examples."

- Update:
The code is updated to have all ".jpg", ".png" and ".jpeg" categorized as browse_* collection products, and all ".fits" as data_* collection products when processing level is not Ancillary . If processing level is Ancillary , they will be categorized into miscellaneous_* collection products.
